skfda/preprocessing/dim_reduction/projection/_fpca.py

```diff
@@ -1,33 +1,445 @@
+        """Computes the n_components first principal components and saves them.
+
+        The eigenvalues associated with these principal
+        components are also saved. For more details about how it is implemented
+        please view the referenced book, chapter 8.
+
+        In summary, we are performing standard multivariate PCA over
+        :math:`\\frac{1}{\sqrt{N}} \mathbf{X} \mathbf{W}^{1/2}` where :math:`N`
+        is the number of samples in the dataset, :math:`\\mathbf{X}` is the data
+        matrix and :math:`\\mathbf{W}` is the weight matrix (this matrix
+        defines the numerical integration). By default the weight matrix is
+        obtained using the trapezoidal rule.
+
+        Args:
+            X (FDataGrid):
+                the functional data object to be analysed in basis
+                representation
+            y (None, not used):
+                only present for convention of a fit function
+
+        Returns:
+            self (object)
+
+        References:
+            .. [RS05-8-4-1] Ramsay, J., Silverman, B. W. (2005). Discretizing
+            the functions. In *Functional Data Analysis* (p. 161). Springer.
+        """
+
+        # check that the number of components is smaller than the sample size
+        if self.n_components > X.n_samples:
+            raise AttributeError("The sample size must be bigger than the "
+                                 "number of components")
+
+        # check that we do not exceed limits for n_components as it should
+        # be smaller than the number of attributes of the funcional data object
+        if self.n_components > X.data_matrix.shape[1]:
+            raise AttributeError("The number of components should be "
+                                 "smaller than the number of discretization "
+                                 "points of the functional data object.")
+
+        # data matrix initialization
+        fd_data = np.squeeze(X.data_matrix)
+
+        # get the number of samples and the number of points of descretization
+        n_samples, n_points_discretization = fd_data.shape
+
+        # if centering is True then subtract the mean function to each function
+        # in FDataBasis
+        if self.centering:
+            meanfd = X.mean()
+            # consider moving these lines to FDataBasis as a centering function
+            # subtract from each row the mean coefficient matrix
+            fd_data -= np.squeeze(meanfd.data_matrix)
+
+        # establish weights for each point of discretization
+        if not self.weights:
+            # sample_points is a list with one array in the 1D case
+            # in trapezoidal rule, suppose \deltax_k = x_k - x_{k-1}, the weight
+            # vector is as follows: [\deltax_1/2, \deltax_1/2 + \deltax_2/2,
+            # \deltax_2/2 + \deltax_3/2, ... , \deltax_n/2]
+            differences = np.diff(X.sample_points[0])
+            self.weights = [sum(differences[i:i + 2]) / 2 for i in
```

Comment:
Vectorize this.

Reply:
Not sure why we should vectorize this as the sample points usually are small?

Reply:
Because it is trivial and more simple:

```
differences = np.concatenate(((0,), differences , (0,)))
self.weights = (differences [:-1] + differences [1:]) / 2
```

Reply:
Done